File: i18ncenter-js/src/translator.ts
import { I18nCenterClient } from './client';
import { DeploymentStage, TranslationData } from './types';

/**
 * Get nested value from an object using dot notation path.
 * Returns the raw value (string or object) without coercion.
 */
function getNestedValue(obj: any, path: string): unknown {
  const keys = path.split('.');
  let current: unknown = obj;

  for (const key of keys) {
    if (current == null || typeof current !== 'object') {
      return null;
    }
    current = (current as Record<string, unknown>)[key];
  }

  return current ?? null;
}

/**
 * Translation function options
 */
export interface TranslateOptions {
  /** Locale override */
  locale?: string;
  /** Stage override */
  stage?: DeploymentStage;
  /** Default value if translation not found */
  defaultValue?: string;
  /** Replace template variables (e.g., {name} -> value) */
  variables?: Record<string, string | number>;
}

/**
 * Create a translation function (t) for a specific component
 */
export function createTranslator(
  client: I18nCenterClient,
  applicationCode: string,
  componentCode: string,
  defaultLocale?: string,
  defaultStage?: DeploymentStage
) {
  // Cache for component translations
  let cachedTranslation: TranslationData | null = null;
  let cachedLocale: string | undefined;
  let cachedStage: DeploymentStage | undefined;

  /**
   * Translation function
   * Usage: t('form.name.label') or t<{b: string}>('a')
   *
   * When T is string (default): returns the translation string, or defaultValue/path on miss.
   * When T is an object type: returns the nested object at that path.
   *
   * Template variable substitution only applies when the resolved value is a string.
   */
  async function t<T = string>(path: string, options?: TranslateOptions): Promise<T> {
    try {
      const locale = options?.locale || defaultLocale || client['config'].defaultLocale;
      const stage = options?.stage || defaultStage || client['config'].defaultStage;

      // Check if we need to reload translation
      if (!cachedTranslation || cachedLocale !== locale || cachedStage !== stage) {
        try {
          cachedTranslation = await client.getTranslation(applicationCode, componentCode, locale, stage);
          cachedLocale = locale;
          cachedStage = stage;
        } catch (error) {
          return (options?.defaultValue ?? path) as unknown as T;
        }
      }

      const value = getNestedValue(cachedTranslation, path);

      if (value === null || value === undefined) {
        return (options?.defaultValue ?? path) as unknown as T;
      }

      // Template variable substitution only for string values
      if (typeof value === 'string' && options?.variables) {
        let result = value;
        for (const [key, val] of Object.entries(options.variables)) {
          result = result.replace(new RegExp(`\\{${key}\\}`, 'g'), String(val));
          result = result.replace(new RegExp(`\\[${key}\\]`, 'g'), String(val));
        }
        return result as unknown as T;
      }

      return value as unknown as T;
    } catch (error) {
      return (options?.defaultValue ?? path) as unknown as T;
    }
  }

  /**
   * Preload translation (useful for SSR)
   */
  async function preload(locale?: string, stage?: DeploymentStage): Promise<void> {
    const loc = locale || defaultLocale || client['config'].defaultLocale;
    const stg = stage || defaultStage || client['config'].defaultStage;
    cachedTranslation = await client.getTranslation(applicationCode, componentCode, loc, stg);
    cachedLocale = loc;
    cachedStage = stg;
  }

  /**
   * Get raw translation data
   */
  async function getRaw(locale?: string, stage?: DeploymentStage): Promise<TranslationData> {
    const loc = locale || defaultLocale || client['config'].defaultLocale;
    const stg = stage || defaultStage || client['config'].defaultStage;
    return await client.getTranslation(applicationCode, componentCode, loc, stg);
  }

  return {
    t,
    preload,
    getRaw,
  };
}

/**
 * Synchronous translation function (requires preloaded data)
 * 
 * @returns Always returns a string. Never throws errors.
 * - Returns the translation value if found
 * - Returns `defaultValue` if provided and translation not found
 * - Returns the `path` itself if translation not found and no defaultValue
 */
export function createSyncTranslator(translationData: TranslationData) {
  return function t<T = string>(path: string, options?: { defaultValue?: string; variables?: Record<string, string | number> }): T {
    try {
      const value = getNestedValue(translationData, path);

      if (value === null || value === undefined) {
        return (options?.defaultValue ?? path) as unknown as T;
      }

      if (typeof value === 'string' && options?.variables) {
        let result = value;
        for (const [key, val] of Object.entries(options.variables)) {
          result = result.replace(new RegExp(`\\{${key}\\}`, 'g'), String(val));
          result = result.replace(new RegExp(`\\[${key}\\]`, 'g'), String(val));
        }
        return result as unknown as T;
      }

      return value as unknown as T;
    } catch (error) {
      return (options?.defaultValue ?? path) as unknown as T;
    }
  };
}

File: i18ncenter-go/translator.go
package i18ncenter

import (
	"fmt"
	"strings"
)

// Translator provides translation functions for a specific component
type Translator struct {
	client         *Client
	applicationCode string
	componentCode  string
	locale         string
	stage          DeploymentStage
	cachedData     TranslationData
}

// NewTranslator creates a new translator for a specific component
// applicationCode is required to differentiate components with the same code in different applications
func NewTranslator(client *Client, applicationCode string, componentCode string, locale string, stage DeploymentStage) *Translator {
	if locale == "" {
		locale = client.config.DefaultLocale
	}
	if stage == "" {
		stage = client.config.DefaultStage
	}

	return &Translator{
		client:          client,
		applicationCode: applicationCode,
		componentCode:   componentCode,
		locale:          locale,
		stage:           stage,
	}
}

// T translates a path to a string value
// Path uses dot notation: "form.name.label" -> translation["form"]["name"]["label"]
// Returns the translated string or the path itself if not found
func (t *Translator) T(path string, defaultValue ...string) (string, error) {
	data, err := t.getData()
	if err != nil {
		return "", err
	}

	value := getNestedValue(data, path)
	if value == nil {
		if len(defaultValue) > 0 {
			return defaultValue[0], nil
		}
		return path, nil
	}

	return fmt.Sprintf("%v", value), nil
}

// Tf translates with template variables
// Supports {variable} and [variable] syntax
func (t *Translator) Tf(path string, variables map[string]interface{}, defaultValue ...string) (string, error) {
	text, err := t.T(path, defaultValue...)
	if err != nil {
		return "", err
	}

	// Replace template variables
	for key, val := range variables {
		text = strings.ReplaceAll(text, fmt.Sprintf("{%s}", key), fmt.Sprintf("%v", val))
		text = strings.ReplaceAll(text, fmt.Sprintf("[%s]", key), fmt.Sprintf("%v", val))
	}

	return text, nil
}

// GetRaw returns the raw translation data for the component
func (t *Translator) GetRaw() (TranslationData, error) {
	return t.getData()
}

// Preload preloads the translation data
func (t *Translator) Preload() error {
	data, err := t.client.GetTranslation(t.applicationCode, t.componentCode, t.locale, t.stage)
	if err != nil {
		return err
	}
	t.cachedData = data
	return nil
}

// getData gets the translation data (from cache or API)
func (t *Translator) getData() (TranslationData, error) {
	if t.cachedData != nil {
		return t.cachedData, nil
	}

	data, err := t.client.GetTranslation(t.applicationCode, t.componentCode, t.locale, t.stage)
	if err != nil {
		return nil, err
	}

	t.cachedData = data
	return data, nil
}

// getNestedValue gets a nested value from a map using dot notation
func getNestedValue(data map[string]interface{}, path string) interface{} {
	keys := strings.Split(path, ".")
	current := interface{}(data)

	for _, key := range keys {
		if currentMap, ok := current.(map[string]interface{}); ok {
			if val, exists := currentMap[key]; exists {
				current = val
			} else {
				return nil
			}
		} else {
			return nil
		}
	}

	// If it's a string, return it
	if str, ok := current.(string); ok {
		return str
	}

	// If it's an object, try to find common fields
	if obj, ok := current.(map[string]interface{}); ok {
		if text, ok := obj["text"].(string); ok {
			return text
		}
		if value, ok := obj["value"].(string); ok {
			return value
		}
		if label, ok := obj["label"].(string); ok {
			return label
		}
		// Return JSON string representation
		return fmt.Sprintf("%v", obj)
	}

	return current
}

